File: golang/src/rudp/rudp.go
package rudp

import (
	"container/list"
	"log"
)

const (
	MAX_PACKAGE_LEN = 0xffff

	_PACAKGE_LEN = 512

	_TYPE_HEARTBEAT = 0
	_TYPE_REQUEST   = 1
	_TYPE_DATA      = 2
)

type stMessage struct {
	data []byte
	id   int
	tick int
}

type stPackageBuffer struct {
	buffer []byte
	sz     int
}

type Rudp struct {
	curr_tick        int
	expired_interval int
	send_id          int
	recv_id_min      int
	recv_id_max      int

	send_queue    *list.List
	recv_queue    *list.List
	history_queue *list.List
	request_queue *list.List
	pkg_queue     *list.List
}

func New() *Rudp {
	return &Rudp{
		curr_tick:        0,
		expired_interval: 10,
		send_id:          0,
		recv_id_min:      0,
		recv_id_max:      0,

		send_queue:    list.New(),
		recv_queue:    list.New(),
		history_queue: list.New(),
		request_queue: list.New(),
		pkg_queue:     list.New(),
	}
}

func new_message(buffer []byte, sz int) *stMessage {
	msg := new(stMessage)
	msg.data = make([]byte, sz)
	copy(msg.data, buffer)
	return msg
}

func new_buffer() *stPackageBuffer {
	return &stPackageBuffer{
		buffer: make([]byte, _PACAKGE_LEN),
		sz:     0,
	}
}

func fill_header(buffer []byte, sz int, id int) int {
	var offset int
	if sz < 128 {
		buffer[0] = byte(sz)
		offset = 1
	} else {
		buffer[0] = byte((sz&0x7f00)>>8) | 0x80
		buffer[1] = byte(sz & 0xff)
		offset = 2
	}

	buffer[offset] = byte((id & 0xff00) >> 8)
	buffer[offset+1] = byte(id & 0xff)

	return offset + 2
}

func get_id(buffer []byte) int {
	return int(buffer[0])*256 + int(buffer[1])
}

func (p *Rudp) insert_message(buffer []byte, sz int, id int) {
	if id < p.recv_id_min { // a past message
		return
	}

	msg := new_message(buffer, sz)
	msg.id = id
	if id > p.recv_id_max || p.recv_queue.Len() == 0 {
		p.recv_queue.PushBack(msg)
		p.recv_id_max = id
	} else {
		for e := p.recv_queue.Front(); e != nil; e = e.Next() {
			tmp := e.Value.(*stMessage)
			if tmp.id == id { // already exists
				return
			} else if tmp.id > id {
				p.recv_queue.InsertBefore(msg, e)
				break
			}
		}
	}
}

func (p *Rudp) insert_request(id int) {
	for e := p.request_queue.Front(); e != nil; e = e.Next() {
		tmp := e.Value.(int)
		if tmp == id {
			return
		} else if tmp > id {
			p.request_queue.InsertBefore(id, e)
			return
		}
	}
	p.request_queue.PushBack(id)
}

func (p *Rudp) pack_message(msg *stMessage, pkgBuffer *stPackageBuffer) {
	dataLen := len(msg.data)
	var pkgLen int
	if dataLen < 128 {
		pkgLen = dataLen + 3
	} else {
		pkgLen = dataLen + 4
	}
	if _PACAKGE_LEN-pkgBuffer.sz < pkgLen {
		p.pkg_queue.PushBack(pkgBuffer)
		pkgBuffer = new_buffer()
	}

	headLen := fill_header(pkgBuffer.buffer[pkgBuffer.sz:], dataLen+_TYPE_DATA, msg.id)
	pkgBuffer.sz += headLen

	copy(pkgBuffer.buffer[pkgBuffer.sz:pkgBuffer.sz+dataLen], msg.data)
	pkgBuffer.sz += dataLen
}

func (p *Rudp) pack_request(pkgBuffer *stPackageBuffer, id int) {
	if _PACAKGE_LEN-pkgBuffer.sz < 3 {
		p.pkg_queue.PushBack(pkgBuffer)
		pkgBuffer = new_buffer()
	}

	pkgBuffer.sz += fill_header(pkgBuffer.buffer[pkgBuffer.sz:], _TYPE_REQUEST, id)
}

func (p *Rudp) pack_heartbeat(pkgBuffer *stPackageBuffer) {
	if pkgBuffer.sz == _PACAKGE_LEN {
		p.pkg_queue.PushBack(pkgBuffer)
		pkgBuffer = new_buffer()
	}
	pkgBuffer.buffer[pkgBuffer.sz] = byte(_TYPE_HEARTBEAT)
	pkgBuffer.sz++
}

func (p *Rudp) unpack(data []byte, sz int) {
	for sz > 0 {
		length := int(data[0])
		if length < 128 {
			data = data[1:]
			sz -= 1
		} else {
			length = (length*256 + int(data[1])) & 0x7fff
			data = data[2:]
			sz -= 2
		}

		switch length {
		case _TYPE_HEARTBEAT:
		case _TYPE_REQUEST:
			p.insert_request(get_id(data))
			sz -= 2
			data = data[2:]
		default:
			length -= _TYPE_DATA
			if sz < length+2 {
				// error
				log.Println("unpack error, length =", length, ", sz =", sz)
				return
			} else {
				id := get_id(data)
				p.insert_message(data[2:], length, id)
				sz -= length + 2
				data = data[length+2:]
			}
		}
	}
}

func (p *Rudp) clear_expired() {
	var next *list.Element
	for e := p.history_queue.Front(); e != nil; e = next {
		next = e.Next()
		tmp := e.Value.(*stMessage)
		if tmp.tick+p.expired_interval < p.curr_tick {
			p.history_queue.Remove(e)
		} else {
			break
		}
	}
}

func (p *Rudp) request_missing(pkgBuffer *stPackageBuffer) {
	id := p.recv_id_min
	for e := p.recv_queue.Front(); e != nil; e = e.Next() {
		tmp := e.Value.(*stMessage)
		if tmp.id > id {
			for i := id; i < tmp.id; i++ {
				p.pack_request(pkgBuffer, i)
			}
		}
		id = tmp.id + 1
	}
}

func (p *Rudp) reply_request(pkgBuffer *stPackageBuffer) {
	for e := p.request_queue.Front(); e != nil; e = e.Next() {
		id := e.Value.(int)
		for ee := p.history_queue.Front(); ee != nil; ee = ee.Next() {
			tmp := ee.Value.(*stMessage)
			if id < tmp.id { // already expired

			} else if id == tmp.id {
				p.pack_message(tmp, pkgBuffer)
			}
		}
	}
	p.request_queue.Init()
}

func (p *Rudp) send_message(pkgBuffer *stPackageBuffer) {
	for e := p.send_queue.Front(); e != nil; e = e.Next() {
		p.pack_message(e.Value.(*stMessage), pkgBuffer)
	}

	p.history_queue.PushBackList(p.send_queue)
	p.send_queue.Init()
}

func (p *Rudp) gen_package() {
	pkgBuffer := new_buffer()

	p.request_missing(pkgBuffer)
	p.reply_request(pkgBuffer)
	p.send_message(pkgBuffer)

	if pkgBuffer.sz == 0 {
		p.pack_heartbeat(pkgBuffer)
	}
	p.pkg_queue.PushBack(pkgBuffer)
}

func (p *Rudp) Send(data []byte, sz int) {
	msg := new_message(data, sz)
	msg.id = p.send_id
	p.send_id++
	msg.tick = p.curr_tick

	p.send_queue.PushBack(msg)
}

func (p *Rudp) Recv() []byte {
	if p.recv_queue.Len() == 0 {
		return nil
	}

	e := p.recv_queue.Front()
	msg := e.Value.(*stMessage)
	if msg.id != p.recv_id_min {
		return nil
	}

	p.recv_id_min++
	p.recv_queue.Remove(e)

	return msg.data
}

func (p *Rudp) Update(data []byte, sz int) *list.List {
	p.curr_tick++
	p.pkg_queue.Init()
	if sz > 0 {
		p.unpack(data, sz)
	}
	p.gen_package()

	l := list.New()
	for e := p.pkg_queue.Front(); e != nil; e = e.Next() {
		tmp := e.Value.(*stPackageBuffer)
		l.PushBack(tmp.buffer[:tmp.sz])
	}

	return l
}

File: c#/rudp.cs
﻿using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace rudp {
    class PackageBuffer {
        public const int PACKAGE_LEN = 512;
        public byte[] buffer;
        public int len;

        public PackageBuffer() {
            buffer = new byte[PACKAGE_LEN];
            len = 0;
        }
    }

    class Rudp {
        const int TYPE_HEARTBEAT = 0;
        const int TYPE_REQUEST = 1;
        const int TYPE_DATA = 2;

        class Message {
            public byte[] data;
            public int id;
            public int tick;
        }

        private int curr_tick;
        private int expired_interval;
        private int send_id;
        private int recv_id_min;
        private int recv_id_max;

        private List<Message> send_queue = new List<Message>();
        private List<Message> recv_queue = new List<Message>();
        private List<Message> history_queue = new List<Message>();
        private List<int> request_queue = new List<int>();
        
        private List<PackageBuffer> pkg_queue = new List<PackageBuffer>();


        public Rudp() {
            curr_tick = 0;
            expired_interval = 10;
            send_id = 0;
            recv_id_min = recv_id_max = 0;
        }

        private Message newMessage(byte[] buffer, int offset, int len) {
            Message msg = new Message();
            msg.id = 0;
            msg.tick = 0;
            msg.data = new byte[len];
            Array.Copy(buffer, offset, msg.data, 0, len);

            return msg;
        }

        private int fillHeader(byte[] buffer, int offset, int len, int id) {
            if (len < 128) {
                buffer[offset] = (byte)len;
                offset += 1;
            } else {
                buffer[offset] = (byte)((len & 0x7f00) >> 8);
                offset += 1;
                buffer[offset] = (byte)(len & 0xff);
                offset += 1;
            }
            buffer[offset] = (byte)((id & 0xff00) >> 8);
            offset += 1;
            buffer[offset] = (byte)(id & 0xff);
            offset += 1;

            return offset;
        }

        private int getId(byte[] buffer, int offset) {
            return buffer[offset] * 256 + buffer[offset + 1];
        }

        private void insertMessage(byte[] buffer, int offset, int len, int id) {
            if (id < recv_id_min) { // a past message
                return;
            }

            Message msg = newMessage(buffer, offset, len);
            msg.id = id;
            if (id > recv_id_max || recv_queue.Count == 0) {
                recv_queue.Add(msg);
                recv_id_max = id;
            } else {
                for (int i = 0; i < recv_queue.Count; ++i) {
                    if (recv_queue[i].id == id) {   // already received
                        return;
                    } else if (recv_queue[i].id > id) { // insert here
                        recv_queue.Insert(i, msg);
                        break;
                    }
                }
            }
        }

        private void insertRequest(int id) {
            int index = -1;
            for (int i = 0; i < request_queue.Count; ++i) {
                if (request_queue[i] == id) {
                    return;
                } else if (request_queue[i] > id) {
                    index = i;
                }
            }
            if (index == -1) {
                request_queue.Add(id);
            } else {
                request_queue.Insert(index, id);
            }
        }

        private void packMessage(Message msg, PackageBuffer pkgBuffer) {
            int lenData = msg.data.Length;
            int lenPackage = (lenData < 128) ? lenData + 3 : lenData + 4;

            if ((PackageBuffer.PACKAGE_LEN - pkgBuffer.len) < lenPackage) {
                pkg_queue.Add(pkgBuffer);
                pkgBuffer = new PackageBuffer();
            }

            pkgBuffer.len = fillHeader(pkgBuffer.buffer, pkgBuffer.len, lenData + TYPE_DATA, msg.id);
            Array.Copy(msg.data, 0, pkgBuffer.buffer, pkgBuffer.len, lenData);
            pkgBuffer.len += lenData;
        }

        private void packRequest(PackageBuffer pkgBuffer, int id) {
            if ((PackageBuffer.PACKAGE_LEN - pkgBuffer.len) < 3) {
                pkg_queue.Add(pkgBuffer);
                pkgBuffer = new PackageBuffer();
            }

            pkgBuffer.len = fillHeader(pkgBuffer.buffer, pkgBuffer.len, TYPE_REQUEST, id);
        }

        private void packHeartbeat(PackageBuffer pkgBuffer) {
            if (pkgBuffer.len == PackageBuffer.PACKAGE_LEN) {
                pkg_queue.Add(pkgBuffer);
                pkgBuffer = new PackageBuffer();
            }
            pkgBuffer.buffer[pkgBuffer.len++] = (byte)TYPE_HEARTBEAT;
        }

        private void unpack(byte[] data, int len) {
            int offset = 0;
            while (offset < len) {
                int tag = data[0];
                if (tag > 127) {
                    tag = (tag * 256 + data[1]) & 0x7fff;
                    offset += 2;
                } else {
                    offset += 1;
                }
                switch (tag) {
                    case TYPE_HEARTBEAT:
                        break;
                    case TYPE_REQUEST:
                        insertRequest(getId(data, offset));
                        break;
                    default:
                        int dataLen = tag - TYPE_DATA;
                        if ((offset + 2 + dataLen) > len) {
                            // error
                            Console.Out.Write("data error\n");
                            return;
                        }
                        insertMessage(data, offset + 2, dataLen, getId(data, offset));
                        offset += 2 + dataLen;
                        break;
                }
            }
        }

        private void clearExpired() {
            int index = -1;
            for (int i = 0; i < history_queue.Count; ++i) {
                Message msg = history_queue[i];
                if ((msg.tick + expired_interval) < curr_tick) {    // expired, remove all msg before this(include this)
                    index = i;
                } else {
                    break;
                }
            }
            history_queue.RemoveRange(0, index + 1);
        }

        private void requestMissing(PackageBuffer buffer) {
            int id = recv_id_min;
            for (int i = 0; i < recv_queue.Count; ++i) {
                if (recv_queue[i].id > id) {
                    for (int j = id; j < recv_queue[i].id; j++) {
                        packRequest(buffer, j);
                    }
                }
                id = recv_queue[i].id + 1;
            }
        }

        private void replyRequest(PackageBuffer pkgBuffer) {
            for (int i = 0; i < request_queue.Count; ++i) {
                int id = request_queue[i];
                int index = -1;
                for (int j = 0; j < history_queue.Count; ++j) {
                    if (id < history_queue[j].id) {
                        break;
                    } else if (id == history_queue[j].id) {
                        index = j;
                        break;
                    }
                }
                if (index == -1) {  // already expired

                } else {
                    packMessage(history_queue[index], pkgBuffer);
                }
            }

            request_queue.Clear();
        }

        private void sendMessages(PackageBuffer pkgBuffer) {
            for (int i = 0; i < send_queue.Count; ++i) {
                packMessage(send_queue[i], pkgBuffer);
            }

            history_queue.AddRange(send_queue);
            send_queue.Clear();
        }

        private void genPackages() {
            PackageBuffer pkgBuffer = new PackageBuffer();

            requestMissing(pkgBuffer);
            replyRequest(pkgBuffer);
            sendMessages(pkgBuffer);

            if (pkgBuffer.len == 0) {
                packHeartbeat(pkgBuffer);
            }

            pkg_queue.Add(pkgBuffer);
        }

        public void Send(byte[] data, int len) {
            Message msg = newMessage(data, 0, len);
            msg.id = send_id++;
            msg.tick = curr_tick;

            send_queue.Add(msg);
        }

        public byte[] Recv() {
            if (recv_queue.Count == 0) {
                return null;
            }
            Message msg = recv_queue[0];
            if (msg.id != recv_id_min) {
                return null;
            }
            recv_id_min++;
            recv_queue.RemoveAt(0);
            return msg.data;
        }

        public List<PackageBuffer> Update(byte[] data = null, int len = 0) {
            curr_tick++;
            pkg_queue.Clear();
            if (data != null && len > 0) {
                unpack(data, len);
            }
            genPackages();

            return pkg_queue;
        }
    }
}
